src/vsr/replica.zig

```diff
@@ -2969,6 +2997,72 @@ pub fn ReplicaType(
             }
         }
 
+        fn on_upgrade_timeout(self: *Self) void {
+            assert(self.primary());
+            assert(self.upgrade_timeout.ticking);
+
+            self.upgrade_timeout.reset();
+
+            if (self.upgrade_release) |_| {
+                // Already upgrading.
+                // Normally we chain send-upgrade-to-self via the commit chain.
+                // But there are a couple special cases where we need to restart the chain:
+                // - The request-to-self might have been dropped if the clock is not synchronized.
+                // - Alternatively, if a primary starts a new view, and an upgrade is already in
+                //   progress, it needs to start preparing more upgrades.
+                self.send_request_upgrade_to_self();
+                return;
+            }
+
+            const release_target: ?u16 = release: {
+                var release_target: ?u16 = null;
+                for (self.releases_bundled.const_slice(), 0..) |release, i| {
+                    if (i > 0) assert(release > self.releases_bundled.get(i - 1));
+                    // Ignore old releases.
+                    if (release <= self.release) continue;
+
+                    var release_replicas: usize = 1; // Count ourself.
+                    for (self.upgrade_targets, 0..) |targets_or_null, replica| {
+                        const targets = targets_or_null orelse continue;
+                        assert(replica != self.replica);
+
+                        for (targets.releases.const_slice()) |target_release| {
+                            assert(target_release > self.release);
+
+                            if (target_release == release) {
+                                release_replicas += 1;
+                                break;
+                            }
+                        }
+                    }
+
+                    // A majority quorum (i.e. `max(quorum_commit, quorum_view_change)`) is required
+                    // to ensure that the upgraded cluster can both commit and view-change.
+                    if (release_replicas >= vsr.quorums(self.replica_count).majority) {
+                        release_target = release;
+                    }
```

Comment:
Replicas advertise (in `ping` messages) the list of release versions their binary supports. When the primary sees that enough replicas have a newer version available, it kicks off the upgrade process.

How many is "enough"? The lower bound is `max(quorum_commit, quorum_view_change) == quorum_majority`, since we want the quorum in the new version to be able to both commit and change views.

I am tempted to set the threshold to `self.replica_count`. In practice, not upgrading all replicas together would probably be an operator error which leads to lagging replicas that then need to state sync.

...But I'm worried about the case of: A replica is broken/recovering, and the operator may _need_ to upgrade anyway. (e.g. to roll out a hotfix quickly)